osquery/carver/carver.cpp

```diff
@@ -0,0 +1,281 @@
+      return Status(1, "Error writing bytes to tmp fs");
+    }
+  }
+
+  return Status(0, "Ok");
+};
+
+Status Carver::compress(const std::set<boost::filesystem::path>& paths) {
+  auto arch = archive_write_new();
+  archive_write_set_format_zip(arch);
+  archive_write_set_format_pax_restricted(arch);
+  archive_write_open_filename(arch, archivePath_.string().c_str());
+
+  for (const auto& f : paths) {
+    PlatformFile pFile(f.string(), PF_OPEN_EXISTING);
+
+    auto entry = archive_entry_new();
+    archive_entry_set_pathname(entry, f.string().c_str());
+    archive_entry_set_size(entry, pFile.size());
+    archive_entry_set_filetype(entry, AE_IFREG);
+    archive_entry_set_perm(entry, 0644);
+    archive_write_header(arch, entry);
+
+    // TODO: Chunking or a max file size.
+    std::ifstream in(f.string(), std::ios::binary);
+    std::stringstream buffer;
+    buffer << in.rdbuf();
+    archive_write_data(arch, buffer.str().c_str(), buffer.str().size());
+    in.close();
+    archive_entry_free(entry);
+  }
+
+  archive_write_close(arch);
+  archive_write_free(arch);
+
+  PlatformFile archFile(archivePath_.string(), PF_OPEN_EXISTING);
+  updateCarveValue(carveGuid_, "size", std::to_string(archFile.size()));
+  updateCarveValue(
+      carveGuid_,
+      "sha256",
+      hashFromFile(HashType::HASH_TYPE_SHA256, archivePath_.string()));
+
+  return Status(0, "Ok");
+};
+
+Status Carver::exfil(const boost::filesystem::path& path) {
+  auto startRequest = Request<TLSTransport, JSONSerializer>(startUri_);
+
+  // Perform the start request to get the session id
+  PlatformFile pFile(path.string(), PF_OPEN_EXISTING);
+  auto blkCount =
+      static_cast<int>(ceil(static_cast<double>(pFile.size()) /
+                            static_cast<double>(FLAGS_carver_block_size)));
+  pt::ptree startParams;
+
+  startParams.put<int>("block_count", blkCount);
+  startParams.put<int>("block_size", FLAGS_carver_block_size);
+  startParams.put<size_t>("carve_size", pFile.size());
+  startParams.put<std::string>("carve_id", carveGuid_);
+  startParams.put<std::string>("node_key", getNodeKey("tls"));
+
+  auto status = startRequest.call(startParams);
+
+  // The call succeeded, store the session id used for future posts
+  boost::property_tree::ptree startRecv;
+  status = startRequest.getResponse(startRecv);
+  if (!status.ok()) {
+    return Status(status.getCode(), status.getMessage());
+  }
+
+  auto session_id = startRecv.get("session_id", "");
+  if (session_id.empty()) {
+    return Status(1, "No session_id received from remote endpoint");
+  }
+
+  auto contRequest = Request<TLSTransport, JSONSerializer>(contUri_);
+  for (int i = 0; i < blkCount; i++) {
+    std::vector<char> block(FLAGS_carver_block_size, 0);
+    pFile.read(block.data(), FLAGS_carver_block_size);
+
+    pt::ptree params;
+    params.put<int>("block_id", i);
+    params.put<std::string>("session_id", session_id);
+    params.put<std::string>("data", base64Encode(block.data()));
```

Comment:
This won't do what you expect in most cases. It's going to return a pointer to the first element in the array (a char) and then treat it as a cstring. This means that it will terminate at the first null byte even if the block isn't complete.